Add concurrency safe iterator wrapper ChannelLike

fredrikekre opened the pull request:
Just to show that the idea from https://github.com/JuliaFolds2/ChunkSplitters.jl/issues/49 can be used to improve the (chunked) GreedyScheduler.
The advantages are that the iterator doesn't have to be copied into a Channel, and that the AtomicChunk-thingy is more efficient (but not as general, of course) than a Channel. See some benchmarks in the test.jl file.

fredrikekre committed:
Add ChannelLike iterator wrapper

`ChannelLike` wraps an indexable object such that it can be iterated by
concurrent tasks in a safe manner similar to a `Channel`. This is
instead of `Channel` in the chunked `GreedyScheduler`.

MasonProtter commented:
Very nice idea, I had wanted something like this before.

fredrikekre commented:
Ok, cleaned it up and added some docs. If you prefer to keep it internal we can do that, but it isn't really much code to maintain. On the other hand, perhaps this isn't a package of "generally useful things". Let me know what you think.

Posting the benchmarks here for posterity since I removed the file with them from the commit:

Benchmark:
```
using OhMyThreads: tmapreduce, GreedyScheduler
using BenchmarkTools: @btime

function mc_parallel(N; ntasks, chunksize)
    scheduler = GreedyScheduler(; ntasks, chunksize, split = :consecutive)
    M = tmapreduce(+, 1:N; scheduler) do i
        rand()^2 + rand()^2 < 1.0
    end
    pi = 4 * M / N
    return pi
end

const N = 100_000_000

for chunksize in (10, 1_000, 100_000), ntasks in (1, 2, 4)
    @btime mc_parallel(N; ntasks = $ntasks, chunksize = $chunksize)
end
```

PR:
```
950.227 ms (15 allocations: 944 bytes)
656.752 ms (21 allocations: 1.48 KiB)
562.387 ms (33 allocations: 2.61 KiB)

506.860 ms (15 allocations: 944 bytes)
252.174 ms (21 allocations: 1.48 KiB)
129.447 ms (33 allocations: 2.61 KiB)

499.351 ms (15 allocations: 944 bytes)
249.004 ms (21 allocations: 1.48 KiB)
127.825 ms (33 allocations: 2.61 KiB)
```

master:
```
2.575 s (53 allocations: 439.07 MiB)
3.593 s (59 allocations: 219.72 MiB)
3.346 s (70 allocations: 110.09 MiB)

518.885 ms (46 allocations: 4.06 MiB)
263.877 ms (52 allocations: 4.06 MiB)
141.339 ms (60 allocations: 69.31 KiB)

500.587 ms (41 allocations: 67.19 KiB)
252.641 ms (47 allocations: 67.75 KiB)
127.858 ms (59 allocations: 26.23 KiB)
```

carstenbauer reviewed (commented):
LGTM and I'm fine with documenting this as API. However, I think we need to modify the `GreedyScheduler` [docstring](https://github.com/JuliaFolds2/OhMyThreads.jl/blob/a301d00be34f187a61f42114e5531f15c96ecc65/src/schedulers.jl#L236-L237) which currently mentions `Channel` explicitly. Probably we shouldn't mention the specific implementation of the "storage" at all.

carstenbauer commented:
BTW, could perhapse make sense to mention this in the docs under https://juliafolds2.github.io/OhMyThreads.jl/stable/literate/tls/tls/#Another-safe-way-based-on-Channel.

fredrikekre commented:
Added to changelog and updated the GreedyScheduler docstring.

carstenbauer commented:
Unfortunately, we mention "channel" more times in the docstring in the keyword arguments section. Would be great if you could "fix" this.

fredrikekre commented:
> BTW, could perhapse make sense to mention this in the docs under

Done.

> Unfortunately, we mention "channel" more times in the docstring

Fixed, replaced with "shared workqueue".
